### DISCUSSION TITLE
Custom rule for console output

### DISCUSSION THREAD

**Discussion by dsl101** [eslint/eslint#17786] [Q&A] [answered]
I have a codebase which uses flags for turning on and off console output via configuration (for logging / debugging, etc.). So there are lots of lines like:
```
   conf.logging && console.log('stuff')
```
Obviously we can insert next-line disable above all of those, but it's tedious, and we'd rather not turn off console warnings completely, as we do want them all 'guarded' like that.

I tried to see if `no-restricted-syntax` could help, but I couldn't see a way to specify `CallExpression[callee.object.name='console']` that is _not_ preceded by a logical expression of type &&. Is there a way to define that?

**Comment by DMartens** [comment, marked as answer]
Yes there are multiple ways, e.g.:
- CSS-like: `:not(LogicalExpression[operator="&&"]) > CallExpression[callee.object.name='console']` 
- `parent` property (just a normal property for the query syntax): `CallExpression[callee.object.name='console'][parent.type!="LogicalExpression"][parent.operator!="&&"]`
You can easily try out the queries and experiment on the [typescript-eslint playground](https://typescript-eslint.io/play) (tab: "ESTree" at the top above the AST)

**Reply by dsl101** [reply]
Thanks. I'd tried something like that first one, but couldn't get it to work for some reason (now deleted, so can't compare). But that works nicely, thank you! I had worried it would create false positives but of course the second part is very specific so no need to worry.
